src/transform/src/analysis/equivalences.rs

```diff
@@ -477,12 +477,148 @@ impl EquivalenceClasses {
+                class.clear();
+            }
+        }
+        novel.retain(|c| !c.is_empty());
+        novel
+    }
+
+    /// Derives potentially novel equivalences without regard for minimization.
+    ///
+    /// This is an opportunity to explore equivalences that do not correspond to expression minimization,
+    /// and therefore should not be used in `minimize_once`. They are still potentially important, but
+    /// required additional guardrails to ensure we reach a fixed point.
+    ///
+    /// The implications will be introduced into `self.classes` and will prompt a round of minimization,
+    /// making it somewhat polite to avoid producing outputs that cannot result in novel equivalences.
+    /// For example, before producing a new equivalence, one could check that the involved terms are not
+    /// already present in the same class.
+    fn implications(&self) -> Vec<Vec<MirScalarExpr>> {
+        let mut new_equivalences = Vec::new();
+
+        // If we see `false == IsNull(foo)` we can add the non-null implications of `foo`.
+        let mut non_null = std::collections::BTreeSet::default();
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_false()) {
+                for e in class.iter() {
+                    if let MirScalarExpr::CallUnary {
+                        func: mz_expr::UnaryFunc::IsNull(_),
+                        expr,
+                    } = e
+                    {
+                        expr.non_null_requirements(&mut non_null);
+                    }
+                }
+            }
+        }
+        // If we see `true == foo` we can add the non-null implications of `foo`.
+        // TODO: generalize to arbitrary non-null, non-error literals; at the moment `true == pred` is
+        // an important idiom to identify for how we express predicates.
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_true()) {
+                for expr in class.iter() {
+                    expr.non_null_requirements(&mut non_null);
+                }
+            }
         }
+        // Only keep constraints that are not already known
+        let lit_false = MirScalarExpr::literal_false();
+        let target = self.remap.get(&lit_false);
+        non_null.retain(|c| {
+            let is_null = MirScalarExpr::column(*c).call_is_null();
+            self.remap.get(&is_null) != target
+        });
+        if !non_null.is_empty() {
+            let mut class = Vec::with_capacity(non_null.len() + 1);
+            class.push(MirScalarExpr::literal_false());
+            class.extend(
+                non_null
+                    .into_iter()
+                    .map(|c| MirScalarExpr::column(c).call_is_null()),
+            );
+            new_equivalences.push(class);
+        }
+
+        // If we see records formed from other expressions, we can equate the expressions with
```

Comment:
The same is true of constant arrays and lists. (IIRC the record case is motivated by some customer queries and these refinements may not be. Might be worth leaving a TODO note?)

Reply:
Yes, probably! I know relatively little about these, is the main limitation!

Reply:
Will kick down the road for a future PR.